pyro/infer/autoguide/guides.py

```diff
@@ -182,12 +178,17 @@ def forward(self, *args, **kwargs):
             self._setup_prototype(*args, **kwargs)
 
         # create all plates
-        self.plates = {frame.name: pyro.plate(frame.name, frame.size, dim=frame.dim)
-                       for frame in sorted(self._plates.values())}
+        with poutine.block():
+            self.plates = {frame.name: pyro.plate(frame.name, frame.size, dim=frame.dim)
+                           for frame in sorted(self._plates.values())}
 
         # run slave guides
         result = {}
-        for part in self.parts:
+        for _, part in self.named_modules():
```

Comment:
Would `self.named_modules(recurse=False)` avoid the need for the `isinstance` check below?

Reply:
I think we would want to recurse if we have one of the subcomponents that is itself an `AutoGuideList`? The check below is needed when for instance we have a neural network in one of the autoguides (AutoIAFNormal, for instance) that we do not want to call. Calling any submodule that isn't an `AutoGuide` instance isn't needed, but moreover since their return type will not be a dict, this will cause exceptions.

Reply:
Thanks, that makes sense that we should treat `AutoGuide`s differently from other `nn.Module`s.

But I think we should set `recurse=False` and omit the `if isinstance(part, AutoGuideList)` below.

Reply:
Actually, I think this should simply iterate over `self._modules`. There is no need to recurse.

Reply:
How about simply:
```diff
- for part in self.parts:
+ for part in self:
```